### ISSUE TITLE
Plan: Fixing existential deposit, bloat bonds, locking and slashing issues

### ISSUE THREAD

**Issue by Lezek123** [Joystream/joystream#4162]
#### Related issues:
* https://github.com/Joystream/joystream/issues/3760
  * bloat bond returning case
  * `MUTATION SAFE` issue no longer applies, as we have protection for that on a framework level now
* https://github.com/Joystream/joystream/issues/4002 (partially)
  * the part that applies to bloat bonds
* https://github.com/Joystream/joystream/issues/3653 (fully)
* https://github.com/Joystream/joystream/issues/4033 (fully)
* https://github.com/Joystream/joystream/issues/3509 (fully)

#### Fixing plan:
1. [x] Never use slashing for bloat bonds. Make sure all bloat bonds are always transferred to the corresponding module account and recoverable (also through transfer) from that account.
1. [x] Make sure all static module accounts are initialized with existential deposit at genesis
    * [x] storage treasury
    * [x] content treasury **_(fixed)_**
    * [x] project token treasury
1. [x] Make sure all dynamic bloat bonds are higher than existential deposit (enforce this if possible), because:
  * for dynamic module accounts (for example: channel module accunts), the bloat bond should simultaneously serve as an existential deposit:
      * [x] channel accounts
      * [x] thread accounts (forum)
      * <s>thread accounts (proposals)</s> (those should be replaced with a single `proposals_discussion` module account, as there are no thread deposits and there's no logic in place that would make this replacement infeasible)
  * <s>transaction to recover a bloat bond may fail otherwise (with `Error::ExistentialDeposit`)</s> (this is not really a breaking issue)
1. [x] Make sure dynamic module accounts (for example: channel accounts), use transfer with `ExistentialRequirement::AllowDeath` when returning the bloat bond (ie. on channel deletion)
1. [x] Update tests and benchmarks

#### Related issues:
* https://github.com/Joystream/joystream/issues/3414 (fully)
* https://github.com/Joystream/joystream/issues/3569 (fully)
  * makes sure locked funds used to pay bloat bonds are not recoverable as usable balance
* https://github.com/Joystream/joystream/issues/4126 (fully)

#### Fixing plan:
1. [x] Implement a new mechanism for checking and paying transactions costs
    <s>**_Draft_**</s> **_EXPIRED, see: https://github.com/Joystream/joystream/issues/4162#issuecomment-1213451412)_**
    <s>
    ```rust
    // Describes a single cost (for example: a bloat bond) associated with an extrinsic
    pub struct Cost {
        // Identifiers of locked balances that can be employed for covering the cost
        allowed_locks: Vec<LockIdentifier>
    
        // Amount to be paid
        amount: Balance
    }
    
    // Ensure all given costs can be covered by a given account
    pub fn can_cover_costs<T>(
        account_id: AccountId,
        costs: &[Cost],
        existance_requirement: ExistenceRequirement
    ) {
        let mut locks = balances::Pallet::<T>::locks(account_id);
        let mut usable = balances::Pallet::<T>::usable_balance(account_id);
        let total = balances::Pallet::<T>::total_balance(account_id);
    
        for cost in costs {
            let mut covered = usable;
            let mut lock_index = 0;
            while covered < cost.amount && lock_index < cost.allowed_locks.len() {
                let lock_id = cost.allowed_locks[lock_index];
                let mut lock = locks.iter_mut().find(|l| l.id == lock_id);
                if Some(lock) = lock {
                    let covered_by_lock = min(lock.amount, cost.amount - covered);
                    lock.amount -= covered_by_lock;
                }
                lock_index += 1;
            }
    
            if covered != cost.amount {
                return false;
            }
        }
    
        existance_requirement == false || total - costs.map(|c| c.amount).sum() >= T::ExistentialDeposit::get()
    }
    
    // Returns: (amount_paid_from_usable_balance, amount_paid_from_locked_balance)
    pub fn pay_cost<T>(
        payer: AccountId,
        reciever: AccountId,
        cost: Cost,
        existance_requirement: ExistenceRequirement
    ) -> Result<(Balance, Balance), DispatchError> {
        let mut locks = balances::Pallet::<T>::locks(payer);
        let mut usable = balances::Pallet::<T>::usable_balance(payer);
        let mut covered = min(usable, cost.amount);
        let mut lock_index = 0;
        let usable_used = covered;
    
        while covered < cost.amount && lock_index < cost.allowed_locks.len() {
            let lock_id = cost.allowed_locks[lock_index];
            let mut lock = locks.iter_mut().find(|l| l.id == lock_id);
            if Some(lock) = lock {
                let covered_by_lock = min(lock.amount, cost.amount - covered);
                let locked_balance_left = lock.amount - covered_by_lock;
                if locked_balance_left == Zero::zero() {
                    balances::Pallet::<T>::remove_lock(lock_id, payer);
                } else {
                    balances::Pallet::<T>::set_lock(lock_id, payer, locked_balance_left, lock.reasons);
                }
            }
            lock_index += 1;
        }
    
        ensure!(covered == cost.amount, DispatchError::other("pay_cost: Insufficient funds"));
    
        let locked_used = cost.amount - usable_used;
    
        <balances::Pallet::<T> as Currency<T>>::transfer(
            payer,
            reciever,
            usable_used,
            existance_requirement
        )?;
    
        balances::Pallet::<T>::burn_from(payer, locked_used)?;
    
        ensure!(
            existance_requirement == ExistanceRequirement::AllowDeath ||
                balances::Pallet::<T>::total_balance(payer) >= T::ExistentialDeposit::get(),
            DispatchError::other("pay_cost: Existance requirement violated")
        );
        
        (usable_used, locked_used)
    }
    ```
    </s>
1. [x] Prepare a list of extrinsics which execution involves additional bloat bonds / custom fees / staking costs throought the runtime. Choose which of them should be compatible with invitation lock.
1. [x] In case any changes are needed w.r.t. staking - just adjust the compatible locks.
1. [x] Use the new mechanism (described in `1.`) to handle bloat bonds and other custom extrinsic costs (fees). Those that should be allowed to be covered with `InvitedMemberLock` funds (as determined in `2.`) should use a `Cost` with `allowed_locks: [InvitedMemberLockId]`

    <s>**_Example implementation for `fn create_channel` (draft):_**</s> **_EXPIRED, see: https://github.com/Joystream/joystream/issues/4162#issuecomment-1213451412)_**
    <s>
    ```rust
    /* ... */
    // Storage module returns costs associated with data objects uploading based on provided params
    let upload_costs: Vec<Cost> = <T as Config>::DataObjectStorage::upload_costs(num_objs, total_size);
    // We create additional `Cost` instance for channel bloat bond
    let channel_bloat_bond = Cost {
        amount: Self::channel_state_bloat_bond_value(),
        allowed_locks: T::ChannelStateBloatBondAllowedLocks::get(), // [InvitedMemberLockId]
    }
    // We ensure that sender can cover all the costs without dropping below existential deposit
    ensure!(
        can_cover_costs::<T>(
            sender,
            [upload_costs, vec![channel_bloat_bond]].concat(),
            ExistenceRequirement::KeepAlive
        ),
        Error::<T>::InsufficientBalanceToCreateChannel
    );
    /* ... */
    // Storage module will execute `pay_cost` for data object bloat bond and data size fee internally
    Storage::<T>::create_dynamic_bag(bag_creation_params)?;
    
    // We can use `usable_balance_used` part of the channel state bloat bond `pay_cost` result
    // as `recoverable_bloat_bond`
    let (recoverable_bloat_bond, _) =
        pay_cost::<T>(sender, channel_acc, channel_bloat_bond, ExistenceRequirement::KeepAlive)?;
    let channel: Channel<T> = ChannelRecord {
        /* ... */
        channel_state_bloat_bond: recoverable_bloat_bond,
    };
    ```
    </s>
1. [x] Update tests and benchmarks

### 3. Address invalid usages of `slash`:
We are using `slash` throught the codebase to handle payments (`slash`

* `deposit_creating`) and transfers to council/wg budget. After some investigation it became obvious that those are not valid use-cases for `Balances::<T>::slash` method, as this generates numerous issues:
* `slash` is indifferent to locked / reserved balance, which shouldn't be allowed to be drawn from in those use-cases
* existential deposit cannot be slashed, even though it's part of the `usable_balance` (which is usually checked before `slash`)
* `slash` is infallible, which means even if we fail to slash the desired amount the extrinsic execution will continue (and the extrinsic will potentially succeed)

#### Related issues:
* https://github.com/Joystream/joystream/issues/4002 (the part that applies to invalid usages of `slash`)
* https://github.com/Joystream/joystream/issues/4105 (full scope)

#### Fixing plan:
1. [x] Use `transfer` instead of `slash` + `deposit_creating` whenever possible, because:
  * it fails unless the entire amount is transferred
  * we can explicitly provide `ExistenceRequirement` based on the use-case
  * it's sensitive to locks and reserved balance
1. [x] If burning is unavoidable (for example, payment involves a platform fee), use the following, common mechanism:
**_Draft:_**
     ```rust
     pub fn can_burn<T>(account: AccountId, amount: Balance, existance_requirement: ExistenceRequirement) -> bool {
         let allowed_withdraw_consequences = match existance_requirement {
             ExistenceRequirement::KeepAlive => vec![WithdrawConsequence::Success],
             ExistenceRequirement::AllowDeath => vec![WithdrawConsequence::Success, WithdrawConsequence::ReducedToZero],
         }
         allowed_withdraw_consequences.contains(balances::Pallet::<T>::can_withdraw(account, amount))
     }

    pub fn safe_burn<T>(account: AccountId, amount: Balance, existance_requirement: ExistanceRequirement) -> DispatchResult {
        // Contrary to `slash`, the `burn_from` method:
        // - fails unless the entire amount is burned
        // - is sensitive to locks and reserved balance
        // - is capable of burning the existential deposit
        let actual = balances::Pallet::<T>burn_from(account, amount)?;
        // We can then additionally check if ExistanceRequirement was satisfied
        ensure!(
            existance_requirement == ExistanceRequirement::AllowDeath ||
                balances::Pallet::<T>::total_balance(account) >= T::ExistentialDeposit::get(),
            DispatchError::other("safe_burn: Existance requirement violated")
        );
        Ok(())
    }
    ```
1. [x] Update tests and benchmarks

## Part 4: Testing final assumptions
* All staking/locking actions:
  * [x] are invoked through `StakingHandler`
  * [x] are sensitive to all locks (including vesting lock), except those that are explicitly compatible with given staking purpose
 (`is_account_free_of_conflicting_stakes` is always checked)
  * [x] are sensitive to reseved balance
* All balance transfers:
  * [x] use `ExistenceRequirement::KeepAlive` unless we're explicitly allowing source account to be removed (ie. during channel bloat bond retrieval)
  * [x] are invoked in a result-sensitive way (`transfer(...)?;`)
  * [x] are not expected to "never fail"
* All burning:
  * [x] is done through the mechanism described in `3.2.` (in a result-sensitive way)
      * True expect for `burn_account_tokens`, which can allow burning any funds (locked / reserved etc.) 
* Bloat bonds:
  * [x] are payable with funds locked via `InvitedMemberLock` (unless explicitly excluded from that rule)
  * [x] are not payable with any other locked funds
  * [x] are not payable with reserved funds
  * <s>are returned only in part that was paid using non-locked balance</s>
  * [x] are returned to the original account that deposited the bloat bond
* All module accounts:
  * [x] are initialized with existential deposit either through gensis config or bloat bond

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1202132419573087/1202782602300825) by [Unito](https://www.unito.io)

**Comment by Lezek123** [comment]
## Correction for Part 2:

### Bad assumptions

In the original implementation plan for **_Part 2_** I made a few bad assumptions, ie.:
- I didn't take into account that `lock.amount` can exceed user's free balance in some cases,
- I didn't consider providing users the ability to recover the locked funds that were used to pay the bloat bond - I assumed that burning those funds and making them unrecoverable is the simplest and safest solution (which is arguably not the case)

Because of that I was also stuck with solutions that either:
- Didn't handle existance deposits for submodule accounts correctly
- Didn't provide a valid way of calculating "withdrawable" channel account balance (which should be: `channel_account_balance - bloat_bond_deposited`)
- Were quite complex

### New implementation plan

In the new implementation plan I'm making the following adjustments:
- Paying a bloat bond with locked funds will not modify `lock.amount` of any of the account's locks in any way. Free balance of the the payer will be adjusted with `Balances::<T>::make_free_balance_be` and the balance of the module account will be increased with `Balances::<T>::deposit_creating`. There will be multiple checks done at each step to make sure the balances are modified exactly as expected.
- Locked funds used to pay the bloat bonds will be recoverable in full, but only to the same account that they were paid from (which will keep being subject to the same locks)

New algorithm for `Cost`s:
- a given `Cost` is payable by an `account` only if `account.free_balance - highest_disallowed_lock >= Cost.amount`, where `highest_disallowed_lock` is the highest of `Balances::<T>::locks(account)`, with the exception of the locks that belong to `Cost.allowed_locks`.
- a set of `Cost`s is payable by an `account` if the above holds true for each `Cost` in the set, assuming we reduce `free_balance` by `Cost.amount` after each iteration and `free_balance` never drops below `0`

**Comment by bedeho** [comment]
> Never use slashing for bloat bonds. Make sure all bloat bonds are always transferred to the corresponding module account and recoverable (also through transfer) from that account.

I am not sure this is good actually, so lets take original set of cited problems with slashing

https://github.com/Joystream/joystream/issues/4033#issue-1305193439

- "We end up impacting issuance, which should be avoided when possible.": true, not the worst thing ever, but true.
- "Bloat bonds are often needed to protect against existential deposit cleanup mechanism": true, but why does that mean we can't use slashing?
- "Unless there is appropriate balance check, slashing allows us to ignore locks & reservations": true, but we cannot avoid this even using transfers or anything else.

you add about benefit of `transfers`

- "it fails unless the entire amount is transferred": I thought slash did this also, but either way, correct balance check is unavoidable, and it should cover this.
- "we can explicitly provide ExistenceRequirement based on the use-case": same?
- "it's sensitive to locks and reserved balance": yes, but not in the way we need.

My main point is that slashing seems to actually be required, because transfers will block on locking reasons that are not aligned with ours. For example, invitation locks are a totally special kind of lock that will be arbitarily allowed or not across a variety of extrinsics.

# Bloat bond fund leaks

Seems to me that there is no part of this that blocks leaking funds through bloat bonds, is there? That would require that recovery can only happen back to a specific account, to block account swapping. So this issue

https://github.com/Joystream/joystream/issues/3569

We probably also would need to do this

> We need to probably introduce the actual amount of bloat bonds used for a given bond, on the relevant object, otherwise, if the bloat bond magnitude is changed (in particular increased) in the interim, then the treasury may become insolvent.

so for example adding `ChannelRecord::only_recover_deletion_prize_to_account: Option<AccountId>`.

This feels very heavy, perhaps not worth fixing now, but we need to atleast decide so explicitly somewhere.

**Comment by bedeho** [comment]
LOL, ok just saw you did comment on fixing the leaks.

So I guess I am just asking: perhaps we should not bother? bloat bonds are not big, its a really cumbersome attack, and it can be solved later without a migration I believe.

**Comment by Lezek123** [comment]
> My main point is that slashing seems to actually be required, because transfers will block on locking reasons that are not aligned with ours. For example, invitation locks are a totally special kind of lock that will be arbitarily allowed or not across a variety of extrinsics.

It may be ok to do slashing if we take extra care with checking pre and post-conditions, but even if we want to draw from the locked balance, there are other methods to do so, for example `balances::withdraw` or `balances::make_free_balance_be`, which at least never touch the reserved balance.

> So I guess I am just asking: perhaps we should not bother?

I'm almost done with implementing this part and I think another benefit of doing this is that it'll also standarize the implementation for bloat bond handling throughout the runtime, so that we can easily make changes in the future.

**Comment by bedeho** [comment]
> I'm almost done with implementing this part and I think another benefit of doing this is that it'll also standarize the implementation for bloat bond handling throughout the runtime, so that we can easily make changes in the future.

But this touches tons of pallets, right? tons of tests, bencmarks, or is this just in one first pallet you have done this?

**Comment by Lezek123** [comment]
From my point of view this is a small proportion of the work. I currently finished the bloat bond rework for content and storage pallet, and this included (among others):
- creating a shared mechanism for paying fees with invitation lock (and optionally some other locks, since it's now configurable)
- implementing this mechanism in content and storage pallet
- adjustments to tests, this was mainly focused on content tests, where I changed the default channel bloat bond from `0` to a bond that can actually serve as existential deposit and then removed existing workarounds where we needed to "manually" initialize channel account with existential deposit,
- adding new tests for different locks scenarios (creating/updating channel w/ invitation lock, w/ a non compatible lock and with both compatible and non-compatible lock)

So the fact that bloat bonds are going to be only recoverable to the same account doesn't really add that much into this. I just created a `RepayableBloatBond` struct with `account`, `amount` and a `repay()` method and it can be reused for any bloat bond. I guess I can add one additional test/assertion for `channel_deletion` / `video_deletion` just to be 100% sure it works as expected and then do the same for forum, proposals discussions etc. where I need to make quite a few changes anyway to support paying the bloat bonds with invitation lock etc.
